installed capacity with initial facility list capability

gwenchee opened the pull request:
This PR addresses https://github.com/arfc/d3ploy/issues/252 and https://github.com/arfc/d3ploy/issues/250 and https://github.com/arfc/d3ploy/issues/241

gwenchee committed:
installed cap now works

gwenchee committed:
readme

gwenchee committed:
some documentation

pep8speaks commented:
Hello @gwenchee! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`d3ploy/demand_driven_deployment_inst.py`](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py):

> [Line 230:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py#L230): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)
> [Line 246:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py#L246): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 249:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py#L249): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 260:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py#L260): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 275:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/demand_driven_deployment_inst.py#L275): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)

* In the file [`d3ploy/supply_driven_deployment_inst.py`](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py):

> [Line 220:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py#L220): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)
> [Line 238:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py#L238): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 241:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py#L241): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 252:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py#L252): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 266:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/d3ploy/supply_driven_deployment_inst.py#L266): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)

* In the file [`tests/integration_tests/installedcap_test.py`](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py):

> [Line 38:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L38): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 39:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 47:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 52:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 56:1](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L56): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 87:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L87): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 95:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L95): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 96:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 110:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L110): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 111:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L111): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 143:1](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L143): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 174:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 182:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L182): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 183:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L183): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 197:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 198:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L198): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 204:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L204): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 233:1](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L233): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 244:80](https://github.com/arfc/d3ploy/blob/b4905c6ed52445150118e3beeae5ceb260839707/tests/integration_tests/installedcap_test.py#L244): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)



##### Comment last updated at 2019-05-08 00:58:32 UTC

gwenchee committed:
pep8

katyhuff commented on README.md:
```diff
@@ -73,12 +73,14 @@ DemandDrivenDeploymentInst:
 The user can define the buffer type in the state variable `buffer_type`.
 If the user wants a 20% value of supply higher than demand, they should input '0.2'
 and if the user wants a 100[whatever unit] value of supply higher than demand, they should input '100'. 
+- **installed_cap**: This is a boolean to determine whether deployment is governed by supply of the commodity of installed capacity for that commodity. 
```
There's something circular about this sentence. Does the last of need to be an or?

Do you mean:

 "True if facility deployment is governed by installed capacity. False if deployment is governed by actual commodity supply/generation."

katyhuff commented on README.md:
```diff
@@ -73,12 +73,14 @@ DemandDrivenDeploymentInst:
 The user can define the buffer type in the state variable `buffer_type`.
 If the user wants a 20% value of capacity higher than supply, they should input '0.2'
 and if the user wants a 100[whatever unit] value of capacity higher than supply, they should input '100'. 
+- **installed_cap**: This is a boolean to determine whether deployment is governed by actual capacity of the commodity of installed capacity for that commodity. 
```
same comment as above.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -107,6 +107,14 @@ class DemandDrivenDeploymentInst(Institution):
+    installed_cap = ts.Bool(
+        doc="Indicates whether or not to use installed capacity as the" +
+        "supply rather than the amount of that commodity in the simulation.",
+        tooltip="Boolean to indicate whether or not to use installed" +
```
one should be super clear about which case is true. Incorporate comment about clarity from readme here in the docstring & tooltip. Perhaps: 

"True if facility deployment is governed by installed capacity. False if deployment is governed by actual commodity supply/generation."

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -107,6 +107,14 @@ class DemandDrivenDeploymentInst(Institution):
         default="POWER"
     )
 
+    installed_cap = ts.Bool(
```
I wonder if this variable (and input tag) should really be called "capacity_driven" (where the alternative is "generation_driven" or "supply_driven" (though supply_driven is likely an overloaded term).

gwenchee replied:
perhaps "installed_capacity_driven" would be the ideal terminology

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
                 self.facility_constraintcommod,
                 self.facility_constraintval)
             commod_list = list(self.commodity_dict.keys())
+            for commod in commod_list:
```
Lines 204-206 can be replaced by one line:

```
self.installed_capacity = dict(zip(commod_list, [defaultdict(float)]*len(commod_list)))
```

That is, the defaultdict(float) function already gives 0.0 as the default value no matter what key is requested, so line 206 isn't needed. 
And, because commod_list is an iterable, we can use the pythonic zip function. This will be much faster than the for loop in 204 & 205.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
                     if val2['constraint_commod'] != '0':
                         commod_list.append(val2['constraint_commod'])
-            commod_list = list(set(commod_list))
+            self.commod_list = list(set(commod_list))
```
why are we re-defining the commod_list here? I'm confused by the loop in 207-210 (probably because the keys and vals are named with no meaning, and there's an inner loop that is non-obvious.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
+            for child in self.children:
+                count = 0
+                for key, val in self.commodity_dict.items():
+                    for key2, val2 in val.items():
```
Please give names to these keys and values. Are the keys commods? and key2... are they facilities? 
Something like the following is more readable and understandable:

```
for commod, producer_dict in self.commodity_dict.items():
  for producer, capacity in producer_dict.items():
  ....
```

It would be best to split up the tasks a bit rather than having this lengthy, deep `for` loop.

gwenchee replied:
I updated the code to reflect a more readable loop

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
+                            itscommod = key
+                if count == 0:
+                    raise Exception(
+                        'The {} facility that was added to the initial facility list must be included in facility_commod and facility_pref'.format(
```
This exception is great.. but is the line too long?

gwenchee replied:
I realize that this exception is not required because there is a built in exception in cyclus when using initialfacilitylist

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
+                for key, val in self.commodity_dict.items():
+                    for key2, val2 in val.items():
+                        if key2 == child.prototype:
+                            count = 1
```
Are we sure we don't add to the count here? It seems like the count should ideally take on more than just a value of 1 or 0.... 

That is, I think you're aiming for:

```suggestion
                            count += 1
```

gwenchee replied:
I realize that this exception is not required because there is a built in exception in cyclus when using initialfacilitylist

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -194,22 +201,37 @@ def enter_notify(self):
+                    raise Exception(
+                        'The {} facility that was added to the initial facility list must be included in facility_commod and facility_pref'.format(
+                            child.prototype))
+                self.installed_capacity[itscommod][0] = self.commodity_dict[itscommod][child.prototype]['cap']
```
Shouldn't we be adding to, rather than replacing this value? What if there are many prototypes of this type?
Also, why access the zeroeth element explicitly rather than using that element as a key??

Why isn't this the strategy:
```suggestion
                self.installed_capacity[itscommod][child.prototype] += self.commodity_dict[itscommod][child.prototype]['cap']
```

gwenchee replied:
The `self.installed_capacity[itscommod][0]` refers to the installed capacity for that commodity at time 0.

gwenchee replied:
I edited the name of the key and vals in the loop so that is more clear. The point of doing that loop right before is to find out what commodity that `child` provides.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -277,18 +310,23 @@ def calc_diff(self, commod, time):
 
     def predict_supply(self, commod):
+        if self.installed_cap:
+            input = self.installed_capacity[commod]
```
Input is a built-in function in python.
You can create a variable called input, but you shouldn't. 
Creating such a variable over-rides the input() function, so it's bad practice. 
https://stackoverflow.com/a/20670757/761664

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -219,17 +241,28 @@ def decision(self):
             if diff < 0:
-                deploy_dict = solver.deploy_solver(
-                    self.commodity_supply, self.commodity_dict, commod, diff, time)
+                if self.installed_cap:
```
These lines are unnecessarily complicated.
I recommend creating the following function:

```
def target():
  if self.installed_cap:
    return self.installed_capacity
  else:
    return self.commodity_supply
```

Then, you can turn all of this copy pasted stuff into just one line.

```
deploy_dict = solver.deploy_solver(self.target(), self. commodity_dict, commod, diff, time)
```

gwenchee replied:
I implemented this

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -219,17 +241,28 @@ def decision(self):
                     for i in range(num):
                         self.context.schedule_build(self, proto)
+                # update installed capacity dict
+                for proto, num in deploy_dict.items():
```
I think you can do this easier with a zip. 
And, ideally this should be its own function.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -219,17 +241,28 @@ def decision(self):
+                    self.installed_capacity[commod][time + 1] = \
+                        self.installed_capacity[commod][time] + \
+                        self.commodity_dict[commod][proto]['cap'] * num
+            else:
```
rather than this extra `if diff<0:...  else:` layer.... 

... couldn't the default num (returned by the solver) be 0 for any prototypes that don't need to be deployed?

FlanFlanagan replied:
I agree, seems like this would be the smart course of action.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -277,18 +310,23 @@ def calc_diff(self, commod, time):
+
         if self.calc_method in ['arma', 'ma', 'arch']:
-            supply = CALC_METHODS[self.calc_method](self.commodity_supply[commod],
+            supply = CALC_METHODS[self.calc_method](input,
```
here, just feed it `target()` from the function suggested above.

gwenchee replied:
I implemented this

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -277,18 +310,23 @@ def calc_diff(self, commod, time):
                                                     back_steps=self.back_steps)
         elif self.calc_method in ['poly', 'exp_smoothing', 'holt_winters', 'fft']:
-            supply = CALC_METHODS[self.calc_method](self.commodity_supply[commod],
+            supply = CALC_METHODS[self.calc_method](input,
```
here, just feed it target() from the function suggested above.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -277,18 +310,23 @@ def calc_diff(self, commod, time):
+            supply = CALC_METHODS[self.calc_method](input,
                                                     back_steps=self.back_steps,
                                                     degree=self.degree)
         elif self.calc_method in ['sw_seasonal']:
```
There is a better way than all of these if/elif layers.

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -277,18 +310,23 @@ def calc_diff(self, commod, time):
         elif self.calc_method in ['sw_seasonal']:
             supply = CALC_METHODS[self.calc_method](
-                self.commodity_supply[commod], period=self.degree)
+                input, period=self.degree)
```
here, feed in target() from the function suggested above (rather than input).

katyhuff commented on d3ploy/supply_driven_deployment_inst.py:
```diff
@@ -98,6 +98,14 @@ class SupplyDrivenDeploymentInst(Institution):
         default=False)
 
+    installed_cap = ts.Bool(
+        doc="Indicates whether or not to use installed capacity as the" +
```
why does this exist in so many places... can this function be inherited from the base class instead?

katyhuff reviewed (changes requested):
Lots of comments. Major comment: Copy-pasting is an indication that one needs a new function or inheritance.

gwenchee commented:
A long time ago, demand driven inst and supply driven inst were torn apart into two different institutions so that one could drive front end deployment and the other back end deployment. But I realize now, that a boolean state variable could have been implemented to toggle between demand driven and supply driven inst. By doing this, there won't be any copy and pasting of code. I will make a new PR to merge them back together and have a toggle instead. @katyhuff

katyhuff commented:
The two institutions can continue to be separate but still inherit from the same base class. https://docs.python.org/2/tutorial/classes.html#multiple-inheritance

gwenchee commented:
Okay I will attempt to do that. @katyhuff

gwenchee committed:
better documentation for installed cap

gwenchee committed:
zip

gwenchee committed:
remove second commod list

gwenchee committed:
better loop names

gwenchee committed:
remove exception

gwenchee committed:
better named loops

gwenchee committed:
bug

gwenchee committed:
target

gwenchee commented:
I addressed most of your comments @katyhuff , the rest of them can be addressed in a different PR where I will attempt to move things into a base class so that repetition is minimized.

gwenchee committed:
installed cap bug

gwenchee committed:
fixed

gwenchee committed:
oops

gwenchee committed:
bug again

gwenchee committed:
supply fix for facilities that leave sim

gwenchee committed:
pep8

FlanFlanagan commented on d3ploy/supply_driven_deployment_inst.py:
```diff
@@ -229,6 +259,11 @@ def decision(self):
                     str(self.commodity_supply[commod][time]) + "\n"
                 with open(commod + ".txt", 'a') as f:
                     f.write(out_text)
+        for child in self.children:
```
I wonder if we use a local variable keep track of the exit times for each children so this loop isn't done each time step. It probably doesn't take too much computation time, but might be a cleaner implementation?

gwenchee replied:
I think a loop at every time step is inevitable because we need to keep track of every single facility that newly enters the simulation.

FlanFlanagan replied:
I was thinking a local variable something like a map of {time:[agent, agent]}

Then just having it check if the current time step is in the keys and decommission the agents in the array. It's probably not that much faster and realistic is more memory intensive. 

I withdraw my comment.

gwenchee commented:
Could we merge this PR then?

gwenchee commented:
@FlanFlanagan @robfairh @katyhuff
